bot/cogs/doc.py

```diff
@@ -0,0 +1,546 @@
+    the given argument otherwise.
+    """
+
+    @staticmethod
+    async def convert(ctx, argument: str):
+        if not argument.isidentifier():
+            raise CogBadArgument(f"`{argument}` is not a valid Python identifier")
+        return argument
+
+
+class DocumentationBaseURL(commands.Converter):
+    """
+    Represents a documentation base URL.
+
+    This converter checks whether the given
+    URL can be reached and requesting it returns
+    a status code of 200. If not, `CogBadArgument`
+    is raised. Otherwise, it simply passes through the given URL.
+    """
+
+    @staticmethod
+    async def convert(ctx, url: str):
+        try:
+            async with ctx.bot.http_session.get(url) as resp:
+                if resp.status != 200:
+                    raise CogBadArgument(
+                        f"HTTP GET on `{url}` returned status `{resp.status_code}`, expected 200"
+                    )
+        except CertificateError:
+            if url.startswith('https'):
+                raise CogBadArgument(
+                    f"Got a `CertificateError` for URL `{url}`. Does it support HTTPS?"
+                )
+            raise CogBadArgument(f"Got a `CertificateError` for URL `{url}`.")
+        except ValueError:
+            raise CogBadArgument(f"`{url}` doesn't look like a valid hostname to me.")
+        except ClientConnectorError:
+            raise CogBadArgument(f"Cannot connect to host with URL `{url}`.")
+        return url
+
+
+class InventoryURL(commands.Converter):
+    """
+    Represents an Intersphinx inventory URL.
+
+    This converter checks whether intersphinx
+    accepts the given inventory URL, and raises
+    `CogBadArgument` if that is not the case.
+    Otherwise, it simply passes through the given URL.
+    """
+
+    @staticmethod
+    async def convert(ctx, url: str):
+        try:
+            intersphinx.fetch_inventory(SphinxConfiguration(), '', url)
+        except AttributeError:
+            raise CogBadArgument(f"Failed to fetch Intersphinx inventory from URL `{url}`.")
+        except ConnectionError:
+            if url.startswith('https'):
+                raise CogBadArgument(
+                    f"Cannot establish a connection to `{url}`. Does it support HTTPS?"
+                )
+            raise CogBadArgument(f"Cannot connect to host with URL `{url}`.")
+        except ValueError:
+            raise CogBadArgument(
+                f"Failed to read Intersphinx inventory from URL `{url}`. "
+                "Are you sure that it's a valid inventory file?"
+            )
+        return url
+
+
+class Doc:
+    def __init__(self, bot):
+        self.base_urls = {}
+        self.bot = bot
+        self.inventories = {}
+        self.headers = {"X-API-KEY": SITE_API_KEY}
+        self.url = f"{SITE_API_URL}/bot/doc"
+
+    async def on_ready(self):
+        await self.refresh_inventory()
+
+    async def refresh_inventory(self):
+        async def update_single(
```

Comment:
Use a private function rather than a local function.

Reply:
This isn't really used anywhere else

Reply:
Still, local functions are considered bad practice. Having it like `async def _update_single(self, ...)` would both make it easier to understand what code is from what function, as well as keeping function locals small, as this function would get created *every time* the function is ran, whereas keeping it on class level would only create it on class creation. This results in a speedup while running, and a bit longer for startup time (both minimal, but small changes make big changes)